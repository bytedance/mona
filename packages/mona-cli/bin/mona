#! /usr/bin/env node

const mona = require('../dist').default;
mona();
