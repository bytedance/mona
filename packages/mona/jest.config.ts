import baseConfig from '../../jestconfig.base';

module.exports = baseConfig;
