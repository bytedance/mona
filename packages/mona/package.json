{
  "name": "@bytedance/mona",
  "version": "0.2.0-alpha.9",
  "description": "Merchant app's developing and building tools",
  "main": "./dist/index",
  "scripts": {
    "build": "yarn clean && ttsc -b",
    "start": "ttsc -b -w",
    "clean": "rm -rf dist",
    "test": "jest"
  },
  "files": [
    "dist"
  ],
  "author": "zhangxiaowei.chris@bytedance.com",
  "license": "MIT",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org"
  },
  "devDependencies": {
    "copy-webpack-plugin": "^10.2.0",
    "react": "^17.0.2",
    "webpack": "^5.58.2",
    "webpack-chain": "^6.5.1"
  },
  "peerDependencies": {
    "react": "^17.0.2"
  },
  "gitHead": "cc5aa5283af595fd7d8a7b4921f54e2797504edd"
}
