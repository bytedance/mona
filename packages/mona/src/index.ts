export { createProjectConfig, ProjectConfig } from './createProjectConfig';
export { createAppConfig, AppConfig } from './createAppConfig';
export { createPageConfig, PageConfig } from './createPageConfig';
export { default as BaseApis } from './BaseApis';
export * from './BaseApis';
export { default as BaseComponents } from './BaseComponents';
export * from './BaseComponents';
export * from './types';
export * from './lifecycle';
