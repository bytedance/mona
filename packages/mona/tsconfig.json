{
  "compilerOptions": {
    "target": "es2017",
    "moduleResolution": "node",
    "strict": true,
    "module": "ESNext",
    "esModuleInterop": true,
    "resolveJsonModule": true,
    "allowSyntheticDefaultImports": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "forceConsistentCasingInFileNames": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "skipLibCheck": true,
    "pretty": true,
    "jsx": "react",
    "sourceMap": true,
    "declaration": true,
    "plugins": [
      {
        "transform": "typescript-transform-paths"
      },
      {
        "transform": "typescript-transform-paths",
        "afterDeclarations": true
      }
    ],
    "outDir": "dist",
    "baseUrl": ".",
    "paths": {
      "@/*": ["src/*"]
    }
  },
  "include": ["src"],
  "exclude": [
    "**/__tests__",
    "**/*.test.ts",
    "**/*.spec.ts",
    "**/*.test.tsx",
    "**/*.spec.tsx"
   ]
}
