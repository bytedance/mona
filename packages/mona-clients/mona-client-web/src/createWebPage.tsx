import React from 'react';
import { isClassComponent } from '@bytedance/mona-shared/dist/reactNode';
import { parseSearch } from '@bytedance/mona-shared/dist/search';

import { LifecycleContext, PageLifecycleGlobalContext, PageLifecycle } from '@bytedance/mona';

export function createPageLifecycle(Component: React.ComponentType<any>) {
  const PageLifecycleContext = new LifecycleContext();
  const pageEntryRef = React.createRef<any>();

  const callLifecycle = (callbackName: PageLifecycle, ...params: any[]) => {
    const cbs = PageLifecycleContext.lifecycle[callbackName] || new Set([]);
    Array.from(cbs).forEach(cb => cb(...params));

    if (pageEntryRef.current?.[callbackName]) {
      return pageEntryRef.current[callbackName](...params);
    }
  };

  const handleResize = (...rest: any[]) => {
    callLifecycle(PageLifecycle.resize, ...rest);
  };
  const handleLoad = (...rest: any[]) => {
    callLifecycle(PageLifecycle.load, ...rest);
  };
  const handleUnload = (...rest: any[]) => {
    callLifecycle(PageLifecycle.unload, ...rest);
  };
  const handleReady = (...rest: any[]) => {
    callLifecycle(PageLifecycle.ready, ...rest);
  };
  const handleVisibilityChange = () => {
    if (document.visibilityState === 'visible') {
      callLifecycle(PageLifecycle.show);
    } else {
      callLifecycle(PageLifecycle.hide);
    }
  };
  document.addEventListener('visibilitychange', handleVisibilityChange);
  // onResize
  window.addEventListener('resize', handleResize);
  class PageWrapper extends React.Component {
    isReachBottom = false;
    componentDidMount() {
      // onLoad
      handleLoad(parseSearch(location.search));
      // onShow & onHide
      if (document.visibilityState === 'visible') {
        callLifecycle(PageLifecycle.show);
      }

      // onReady
      handleReady();
    }
    componentWillUnmount() {
      // onUnload
      handleUnload();
      window.removeEventListener('resize', handleResize);
      document.removeEventListener('visibilitychange', handleVisibilityChange);
    }

    render() {
      if (isClassComponent(Component)) {
        return <Component {...this.props} ref={pageEntryRef}></Component>;
      }
      return <Component {...this.props}></Component>;
    }
  }

  return (props: any) => (
    <PageLifecycleGlobalContext.Provider value={PageLifecycleContext}>
      <PageWrapper {...props} />
    </PageLifecycleGlobalContext.Provider>
  );
}
