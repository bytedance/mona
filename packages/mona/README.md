# @bytedance/mona
对外暴露提供构建时API
## 配置创建
### createProjectConfig
创建项目配置，`mona.config.ts`中使用
#### 例子
```js
// mona.config.ts
import { createProjectConfig } from '@bytedance/mona';
export default createProjectConfig({
  projectName: 'demo',
  input: './src/app.tsx',
  output: 'dist'
})
```

#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| projectName  | 应用/插件名称 | 是 | string | - |
| input  | 入口文件路径 | 否 | string | - |
| output  | 打包后的文件目录路径 | 否 | string | - |
| raw  | 定义函数来修改webpack配置 | 否 | (options: any) => any | - |
| dev  | 用来修改本地服务配置 | 否 | string | { port: 9999 } |

### createAppConfig
创建应用配置，`app.config.ts`中使用
#### 例子
```js
// app.config.ts
import { createAppConfig } from '@bytedance/mona';
export default createAppConfig({
  pages: [
    'pages/Home/index',
    'pages/Info/index',
    'pages/List/index'
  ]
})
```
#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| pages  | 页面路径 | 是 | string[] | - |

默认`pages`的第一项为首页
### createPageConfig
创建页面配置，需在每个页面目录的`page.config.ts`中使用

#### 例子
```js
// page.config.ts
import { createPageConfig } from '@bytedance/mona';
export default createPageConfig({
  navigationBarTitleText: 'page title'
})
```
#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| navigationBarTitleText  | 当前页面标题 | 否 | string | - |
