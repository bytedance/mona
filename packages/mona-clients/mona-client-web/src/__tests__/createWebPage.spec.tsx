import React from 'react';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { usePageEvent, PageLifecycle } from '@bytedance/mona';
import { createPageLifecycle } from '../createWebPage';
// import createApp from '../createApp';
import { configure, mount } from 'enzyme';

import mountTest from '../../../../tests/shared/mountTest';

configure({ adapter: new Adapter() });

describe('createPageLifecycle', () => {
  let launchMock = jest.fn();
  beforeEach(() => {
    launchMock = jest.fn();
  });

  function Page({ name }) {
    usePageEvent(PageLifecycle.load, (...rest) => {
      launchMock(...rest, 'load');
    });
    usePageEvent(PageLifecycle.ready, (...rest) => {
      launchMock(...rest, 'ready');
    });
    usePageEvent(PageLifecycle.unload, (...rest) => {
      launchMock(...rest, 'unload');
    });
    return <>fc组件{name}</>;
  }
  class PageClass extends React.Component<{ name: string }> {
    onLoad(...rest) {
      launchMock(...rest, 'onLoad');
    }
    onReady(...rest) {
      launchMock(...rest, 'onReady');
    }
    onUnload(...rest) {
      launchMock(...rest, 'onUnload');
    }
    render() {
      return <>class组件{this.props.name}</>;
    }
  }
  mountTest(createPageLifecycle(Page));
  mountTest(createPageLifecycle(PageClass));

  it('createPageLifecycle FC lifeCycle', () => {
    const PageComponent = createPageLifecycle(Page);
    const PageIns = mount(<PageComponent name="mona" />);
    expect(PageIns.render()).toMatchSnapshot();

    // PageIns.props
    expect(PageIns.prop('name')).toBe('mona');

    expect(launchMock.mock.calls.length).toBe(2);
    PageIns.unmount();
    // console.log(launchMock.mock.calls);
    expect(launchMock.mock.calls.length).toBe(3);
  });
  it(`createPageLifecycle Class lifeCycle `, () => {
    const PageComponent = createPageLifecycle(PageClass);
    const PageIns = mount(<PageComponent name="mona" />);
    expect(PageIns.render()).toMatchSnapshot();

    // console.log(' PageIns.find("Component")', PageIns.find('Component'));
    expect(PageIns.prop('name')).toBe('mona');

    expect(launchMock.mock.calls.length).toBe(2);

    // expect(PageIns.)
    PageIns.unmount();
    // console.log(launchMock.mock.calls);
    expect(launchMock.mock.calls.length).toBe(3);
  });
});
