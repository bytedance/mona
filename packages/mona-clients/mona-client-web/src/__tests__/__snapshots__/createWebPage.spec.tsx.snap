// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPageLifecycle createPageLifecycle Class lifeCycle  1`] = `
LoadedCheerio {
  "0": Document {
    "children": Array [
      class组件mona,
    ],
    "endIndex": null,
    "next": null,
    "parent": null,
    "prev": null,
    "startIndex": null,
    "type": "root",
  },
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        class组件mona,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`createPageLifecycle createPageLifecycle FC lifeCycle 1`] = `
LoadedCheerio {
  "0": Document {
    "children": Array [
      fc组件mona,
    ],
    "endIndex": null,
    "next": null,
    "parent": null,
    "prev": null,
    "startIndex": null,
    "type": "root",
  },
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        fc组件mona,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
